File: 248-Strobogrammatic Number III/strobogrammatic_number_ii.cpp
class Solution {
public:
    bool ge(string a, string b){
        int lena = a.size();
        int lenb = b.size();
        if(lena > lenb) return true;
        else if(lena < lenb) return false;
        else{
            int i = 0;
            while(i < lena){
                if(a[i] < b[i]) return false;
                else if(a[i] > b[i]) return true;
                else i++;
            }
            return true;
        }
    }
    
    vector<string> dfs(string low, string high, int len, int cur){
        vector<string> res;
        if(cur == 0) return res = {""};
        if(cur == 1) return res = {"0", "1", "8"};
        vector<string> mid = dfs(low, high, len, cur-2);
        for(auto s : mid){
            if(cur != len) res.push_back("0" + s + "0");
            res.push_back("1" + s + "1");
            res.push_back("6" + s + "9");
            res.push_back("8" + s + "8");
            res.push_back("9" + s + "6");
        }
        return res;
    }
    int strobogrammaticInRange(string low, string high) {
        int res = 0;
        int len_l = low.length();
        int len_h = high.length();
        for(int i = min(len_l, len_h); i <= max(len_l, len_h); i ++){
            vector<string> c = dfs(low, high, i, i);
            for(int i = 0; i < c.size(); i ++){
                if(ge(c[i], low) && ge(high, c[i])) res++;
            }
        }
        return res;
    }
};

File: 248-Strobogrammatic Number III/strobogrammatic_number_ii.py
class Solution:
    def dfs(self, low, high, l, target):
        if l == 0:
            return [""]
        if l == 1:
            return ["0", "1", "8"]
        mid = self.dfs(low, high, l-2, target)
        res = []
        for s in mid:
            if l != target:
                res.append("0"+s+"0")
            res.append("1"+s+"1")
            res.append("6"+s+"9")
            res.append("8"+s+"8")
            res.append("9"+s+"6")
        return res
    
    def strobogrammaticInRange(self, low, high):
        """
        :type low: str
        :type high: str
        :rtype: int
        """
        self.minlen = min(len(high), len(low))
        self.maxlen = max(len(high), len(low))
        res = 0
        for l in range(self.minlen, self.maxlen+1):
            c = self.dfs(low, high, l, l)
            for s in c:
                if int(s) >= int(low) and int(s) <= int(high):
                    res += 1
        return res
